### ISSUE TITLE
[BUG] The new secret webhook may cause CI to fail

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#6627]
**Describe the bug**

The CI may fail from time to time.

https://github.com/harvester/harvester/actions/runs/11008796171/job/30578651252

```
make ci log..
...
requested resource (get virtualmachines.meta.k8s.io)
Add/Modify event for harvester-pre-delete: MODIFIED
Starting delete for "harvester-pre-delete" Job
  uninstall: given cascade value: , defaulting to delete propagation background
Starting delete for "harvester-webhook" Service
Starting delete for "harvester-load-balancer-webhook" Service
Starting delete for "harvester-network-webhook" Service
Starting delete for "harvester-node-manager-webhook" Service
Starting delete for "harvester-snapshot-validation-webhook" Service
Starting delete for "harvester" Service
Starting delete for "harvester-network-webhook" Deployment
Starting delete for "harvester-snapshot-validation-webhook" Deployment
Starting delete for "harvester-node-manager-webhook" Deployment
Starting delete for "virt-operator" Deployment
Starting delete for "harvester" Deployment
Starting delete for "harvester-network-controller-manager" Deployment
Starting delete for "harvester-webhook" Deployment
...
Starting delete for "harvester-helpers" ConfigMap
Starting delete for "whereabouts-config" ConfigMap
Starting delete for "vip" ConfigMap

Starting delete for "snapshot-validation-webhook-tls" Secret

Starting delete for "harvester" ServiceAccount
Starting delete for "snapshot-controller" ServiceAccount
Starting delete for "harvester-load-balancer" ServiceAccount
Starting delete for "harvester-load-balancer-webhook" ServiceAccount
Starting delete for "harvester-network-controller" ServiceAccount
Starting delete for "harvester-network-helper" ServiceAccount
Starting delete for "harvester-network-webhook" ServiceAccount
Starting delete for "harvester-node-manager" ServiceAccount
Starting delete for "harvester-node-manager-webhook" ServiceAccount
Starting delete for "kube-vip" ServiceAccount
Starting delete for "kubevirt-operator" ServiceAccount
Starting delete for "snapshot-validation-webhook" ServiceAccount
Starting delete for "harvester-whereabouts" ServiceAccount
Starting delete for "harvester" NetworkPolicy
Starting delete for "backing-image-data-source" NetworkPolicy
Starting delete for "backing-image-manager" NetworkPolicy
Starting delete for "instance-manager" NetworkPolicy
Starting delete for "longhorn-manager" NetworkPolicy
Starting delete for "longhorn-recovery-backend" NetworkPolicy
Starting delete for "longhorn-conversion-webhook" NetworkPolicy
Starting delete for "longhorn-admission-webhook" NetworkPolicy
Starting delete for "cattle-monitoring-prometheus" NetworkPolicy
Starting delete for "kubevirt-cluster-critical" PriorityClass
Starting delete for "kubevirt" KubeVirt
Ignoring delete failure for "kubevirt" kubevirt.io/v1, Kind=KubeVirt: kubevirts.kubevirt.io "kubevirt" not found
Starting delete for "sync-additional-ca" Plan
Starting delete for "harvester-snapshot-validation-webhook" ValidatingWebhookConfiguration
Starting delete for "longhorn" VolumeSnapshotClass
  purge requested for harvester
  [FAILED] in [AfterSuite] - /go/src/github.com/harvester/harvester/tests/framework/dsl/action.go:44 @ 09/24/24 11:26:39.961
[AfterSuite] [FAILED] [31.699 seconds]
[AfterSuite] 
/go/src/github.com/harvester/harvester/tests/integration/api/suite_test.go:150

  [FAILED] Unexpected error:
      <*errors.errorString | 0xc00a32ae20>: 
      failed to uninstall harvester chart: uninstallation completed with 1 error(s): uninstall: Failed to purge the release: Internal error occurred: failed calling webhook "validator.harvesterhci.io": failed to call webhook: Post "https://harvester-webhook.harvester-system.svc:443/v1/webhook/validation?timeout=10s": service "harvester-webhook" not found
      {
          s: "failed to uninstall harvester chart: uninstallation completed with 1 error(s): uninstall: Failed to purge the release: Internal error occurred: failed calling webhook \"validator.harvesterhci.io\": failed to call webhook: Post \"https://harvester-webhook.harvester-system.svc:443/v1/webhook/validation?timeout=10s\": service \"harvester-webhook\" not found",
      }
  occurred
  In [AfterSuite] at: /go/src/github.com/harvester/harvester/tests/framework/dsl/action.go:44 @ 09/24/24 11:26:39.961

  Full Stack Trace
    github.com/harvester/harvester/tests/framework/dsl.MustNotError({0x3c60520?, 0xc00a32ae20?})
    	/go/src/github.com/harvester/harvester/tests/framework/dsl/action.go:44 +0x4b
    github.com/harvester/harvester/tests/integration/api_test.init.func6()
    	/go/src/github.com/harvester/harvester/tests/integration/api/suite_test.go:153 +0x59
------------------------------
...
```

**To Reproduce**
1. The new feature added a webhook on secret deletion https://github.com/harvester/harvester/blob/0bde44193ce1c2b2e2787237a2b31eb2263fc756/pkg/webhook/resources/secret/validator.go#L36
2. The `make ci` will run as following, which inturn deletes all resources from the chart
```
func uninstallHarvesterCharts(ctx context.Context, kubeConfig *restclient.Config) error {
	// uninstall chart
	_, err := helm.UninstallChart(testChartReleaseName, testHarvesterNamespace)
	if err != nil {
		return fmt.Errorf("failed to uninstall harvester chart: %v", err)
	}

// UninstallChart uninstalls the chart.
func UninstallChart(releaseName string, namespace string) (*release.UninstallReleaseResponse, error) {
...
	return action.NewUninstall(helmCfg).Run(releaseName)
}
```
3. `harvester-webhook` is removed at the beginning
4. There happens to have a `secret` resources
```
Starting delete for "snapshot-validation-webhook-tls" Secret
```
5. The webhook will be called but failed

**Expected behavior**

No CI failure.

Follow https://github.com/harvester/harvester/blob/0bde44193ce1c2b2e2787237a2b31eb2263fc756/deploy/charts/harvester/templates/job.yaml#L1 to add a `pre-delete` job which removes the secret webhook

**Support bundle**

**Environment**
 - Harvester ISO version: v1.4.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

**Comment by w13915984028** [comment]
cc @Yu-Jack @bk201

**Comment by Yu-Jack** [comment]
I'm curious why helm uninstallation doesn't delete webhook, so I did some investigation and provided some contexts.

I checked these two actions.

- Succeed - https://github.com/harvester/harvester/actions/runs/11008796171/job/30583327089
- Failed - https://github.com/harvester/harvester/actions/runs/11008796171/job/30578651252

I noticed that helm uninstallation only processed this deletion `Starting delete for "harvester-snapshot-validation-webhook" ValidatingWebhookConfiguration`. However, it processed this deletion `Starting delete for "harvester-webhook" Service`  in the beginning. 

Then, I found out they are managed by different tools.

![image](https://github.com/user-attachments/assets/e6118ed1-7888-4c19-a332-df79f16c7b90)
![image](https://github.com/user-attachments/assets/f2dfa2b9-6b20-4573-b9a7-da0f0bc02921)
![image](https://github.com/user-attachments/assets/f4881e6c-38af-400a-854d-0474effa07ce)


> p.s. We can also check `./deploy/charts/harvester` to find relation.

**Comment by w13915984028** [comment]
@bk201 regarding your question: https://github.com/harvester/harvester/pull/6629#pullrequestreview-2326916250
 @Yu-Jack 

This is a sequence issue, it does not happen 100%.

```
from 
Starting delete for "harvester-webhook" Service
to
Starting delete for "snapshot-validation-webhook-tls" Secret
```
it depends on how fast k8s 

deletes the `harvester-webhook` service and related pods  
VS 
deletes the secret, 

the secret has a `delete webhook`, if `harvester-webhook` is gone, it will report error, otherwise it can pass.

**Comment by bk201** [comment]
The explanation sounds sane to me, thanks!

**Comment by bk201** [comment]
PRs merged for both master and v1.4 branches. We can observe if the issue happens again.

**Comment by w13915984028** [comment]
Observed a new fail on https://github.com/harvester/harvester/actions/runs/11031062806/job/30637017434, I will continue to check it.
